teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+* Provide sample types and controllers demonstrating simple behavior.
+
+## Proposal
+
+Tekton Pipelines will add a new type, `Run`, initially in `tekotn.dev/v1alpha1`.
+
+The `Run` type will take a reference to a user-defined CRD object:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  generateName: run-
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+```
+
+This references an `Example` CRD type defined by the custom task author, an instance of which is named `my-example`.
+
+When a `Run` object is created, Tekton will validate that the `ref` is specified, and that the specified CRD type is defined, using webhook validation.
+
+After that, Tekton Pipelines expects a custom task author to implement a controller for `Run` objects that reference their type (annotated throughout this proposal with the shorthand `Run<Example>`) to take some action, and eventually update its `.status` to signal completion, either successfully or unsuccessfully, using the `conditions` model used by Tekton PipelineRuns and TaskRuns.
+
+### Validation
+
+Custom Task authors can implement webhook validation for CR objects of their provided type (e.g., to validate `Example` object definitions). Validation is optional but recommended, and examples and sample frameworks will demonstrate this functionality.
+
+### Status Reporting
+
+When the `Run<Example>` is validated and created, the Custom Task controller should be notified and begin doing some operation. When the operation begins, the controller should update the `Run`'s `.status.conditions` to report that it's ongoing:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: Unknown
+```
+
+When the operation completes, if it was successful, the condition should report `status: True`, and optionally a brief `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+```
+
+If the operation was _unsuccessful_, the condition can report `status: False`, and optionally a `reason` and human-readable `message`:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: False
+    reason: ExampleFailed
+    message: Oh no bad times
+```
+
+The `Run` type's `.status` will also allow controllers to report other fields, such as `startTime`, `completionTime`, `results` (see below), and arbitrary context-dependent fields the Custom Task author wants to report. A fully-specified `Run` status might look like:
+
+```
+status
+  conditions:
+  - type: Succeeded
+    status: True
+    reason: ExampleComplete
+    message: Yay, good times
+  completionTime: "2020-06-18T11:55:01Z"
+  startTime: "2020-06-18T11:55:01Z"
+  results:
+  - name: first-name
+    value: Bob
+  - name: last-name
+    value: Smith
+  arbitraryField: hello world
+  arbitraryStructuredField:
+    listOfThings: ["a", "b", "c"] 
```

Comment:
I think we figured out in doc comments that we need to support arbitrary structured fields, to be able to express underlying `TaskRunStatus`es, for instance.

I'd like to make it Tekton's responsibility to snapshot the task's spec (and resolve param placeholders!) but I thought you were saying that was a bad idea for some reason I don't remember 😅

Reply:
My comment was more about having a common shape for that vs. who bore the responsibility of doing so.  It might be easier to quickly sync on this in a more favorable medium tomorrow or next week.

Reply:
In any case, we shouldn't prevent arbitrary fields on the status, right ?
It *shouldn't* affect the duct typing system (as as long as it satisfy the contract it's good to go)